Python 3.12: Add 3.12.2

Bill-Sommerfeld opened the pull request:
~~This is a very very very rough first draft; it makes it all the way through `gmake package` and passes most of the tests.~~
~~patches-todo has patches I haven't dealt with yet.~~
patches has a bunch of hacks that need to be cleaned up.

Bill-Sommerfeld committed:
Python 3.12 import: work in progress

mtelka commented:
Just first suggestion: please switch to `components/python/python-312`.  This is to align with other versioned components.  Thank you.

Bill-Sommerfeld committed:
Rename components/python/python312 to components/python/python-312

Bill-Sommerfeld committed:
shared-macros.mk: Add current estimated Python 3.12 EOL date

Bill-Sommerfeld committed:
Clean up module additions

Bill-Sommerfeld committed:
More patch cleanups.

Bill-Sommerfeld commented:
now down to three main loose ends:

 1) `_curses_panel` isn't being built; need to get to the bottom of why configure doesn't think it could work.
 2) `08-py_db.patch` needs rework as the underlying python internals have changed
 3) need to look more closely at `35-20142.patch` and figure out if it's actually been fixed in 3.12 in some different way. 

and then need to go over all the patches and clean up warts (particularly the 00*.patch hacks).

Bill-Sommerfeld committed:
_curses_panel now builds (though the config is a bit sloppy)

Bill-Sommerfeld committed:
python-312: update TODO

Bill-Sommerfeld committed:
Clean up the patches that fixed up ldflags for stdlib modules.

Bill-Sommerfeld committed:
Tests are happy.

mtelka commented:
> Tests are happy.

Congrats!

Bill-Sommerfeld committed:
Tweak 35-20142.patch to apply cleanly

Bill-Sommerfeld commented:
At this point, the one remaining loose end is `08-py_db.patch` which enables MDB python support (including `::pystack`); this is likely to be tricky as the interpreter structures used by this patch have changed significantly since Python 3.9 was released and it's going to require a rewrite.   I don't see lack of this feature as a reason to delay adding Python 3.12 - at this point it would make sense to squash the commits and start review. 

Thoughts?

mtelka commented:
FYI, there is `refresh-patches` make target.

iigs commented:
Thank you for working on python 3.12. I do not have any comments for your recent question.  Is there a reason that openssl1.1 needs to be used?  Please consider changing "USE_OPENSSL11= yes" to 'OPENSSL_VERSION= 3.1" in the Makefile.

mtelka commented:
I would do not bother with `08-py_db.patch` unless it is easily portable from https://github.com/oracle/solaris-userland/blob/master/components/python/python311/patches/08-py_db.patch.  If it is not portable then just add a note somewhere (in `Makefile` likely) so we should look at `solaris-userland` to see whether the debugger support is available for Python 3.12.

Thank you!

Bill-Sommerfeld committed:
Build against OpenSSL 3.1

Bill-Sommerfeld commented:
> Is there a reason that openssl1.1 needs to be used? Please consider changing "USE_OPENSSL11= yes" to 'OPENSSL_VERSION= 3.1" in the Makefile.

Fixed as you suggested.   I started with `python39/Makefile`, so USE_OPENSSL11 was in there all along; there are likely other stale things in the Makefile.

Bill-Sommerfeld committed:
one more file for the openssl change

Bill-Sommerfeld committed:
remove NEWS.d changes that confused Quilt

Bill-Sommerfeld committed:
Refreshed patches with gmake refresh-patch

Bill-Sommerfeld commented:
> FYI, there is `refresh-patches` make target.

Done.

Bill-Sommerfeld committed:
Remove --with-system-ffi CONFIGURE_OPTION (was obsoleted several releases ago)

Bill-Sommerfeld committed:
mdb python support: it builds

Bill-Sommerfeld commented:
> I would not bother with `08-py_db.patch` unless it is easily portable from https://github.com/oracle/solaris-userland/blob/master/components/python/python311/patches/08-py_db.patch. If it is not portable then just add a note somewhere (in `Makefile` likely) so we should look at `solaris-userland` to see whether the debugger support is available for Python 3.12.

The equivalent file in the solaris-userland patch has an odd copyright statement in it that grants rights only to the PSF so I'm not touching it.

I took a closer look at the data structure differences between 3.9 and 3.12 and it didn't look that bad.  I've managed to tweak our version into building; to run it I'll need to build a corresponding `/usr/lib/mdb/proc/amd64/libpython3.12.so` from illumos-gate, which I think can wait for a bit.

mtelka commented:
@Bill-Sommerfeld, please do NOT squash yet.  The history contains valuable information that might be needed during review.

Thank you.

mtelka committed:
Drop python312.license; switch to SPDX license identifier

mtelka commented on components/python/python-312/Makefile:
```diff
@@ -0,0 +1,208 @@
+CONFIGURE_ENV += ac_cv_func_epoll_ctl=no
+CONFIGURE_ENV += ac_cv_func_epoll_create1=no
+
+CONFIGURE_ENV += ac_cv_func_getentropy=no
```
This is similar to line 119 below, but it is commented there.  The linked bug (see below) is closed as resolved (with fix merged) for many years now.  Could you please make sure this one is still needed?  If so, then please add some reasonable comment explaining the reason.

Thank you.

Bill-Sommerfeld replied:
As far as I can tell disabling getentropy is not needed and these lines will be gone in the next push.   Truss on a short python script shows that os.urandom() generates a call into the getrandom() syscall passing a zero flag (which is the desirable fast-path, unlikely to block) rather than calling getentropy().

mtelka replied:
> As far as I can tell disabling getentropy is not needed and these lines will be gone in the next push. Truss on a short python script shows that os.urandom() generates a call into the getrandom() syscall passing a zero flag (which is the desirable fast-path, unlikely to block) rather than calling getentropy().

I assume you trussed with these lines removed... (just to be sure we are on the same page :-))

Bill-Sommerfeld replied:
The code in `Python/bootstrap_hash.c` is structured such that it will only build code to use getentropy if getrandom isn't present:

```
#if defined(HAVE_GETRANDOM) || defined(HAVE_GETRANDOM_SYSCALL)
#define PY_GETRANDOM 1
...
#elif defined(HAVE_GETENTROPY)
#define PY_GETENTROPY 1
...
#endif /* defined(HAVE_GETENTROPY) && !(defined(__sun) && defined(__SVR4)) */
```
Just to be sure I'll test this again with a build with the lines removed.

Bill-Sommerfeld replied:
> Just to be sure I'll test this again with a build with the lines removed.

I've retested.  As expected, removing the `...getentropy=no` lines from the Makefile had no effect on the behavior - `os.urandom()` results in a `getrandom()` syscall with them either present or  removed.;

mtelka committed:
Drop COMPONENT_PUBLISH_ENV

mtelka committed:
Drop PATH from ENV; it is by default there already

mtelka committed:
Drop ENV setting

mtelka committed:
Rename manifests to do not contain $(PYV)

mtelka commented:
@Bill-Sommerfeld, you probably already noticed, but if not, then: https://github.com/Bill-Sommerfeld/oi-userland/pulls

mtelka committed:
PYTHON_VERSION

mtelka committed:
shared-macros.mk: remove extra line

mtelka committed:
Support for versions like 3.12.0a1

mtelka committed:
Remove CONFIGURE_FIX_LIBTOOL_RPATH

Bill-Sommerfeld committed:
Revise ucred patch to set library dependencies in configure.ac

Move illumos additions in configure.ac before the
AC_SUBST([MODULE_BLOCK]) which is described as "must come after last
PY_STDLIB_MOD()".

Build ucred with -Werror so we catch function deprecations early.

Bill-Sommerfeld committed:
Revise dlpi patch to set library dependencies in configure.ac

Fix all compile-time warnings.

Move illumos additions in configure.ac before the
AC_SUBST([MODULE_BLOCK]) which is described as "must come after last
PY_STDLIB_MOD()".

Build dlpi module with -Werror so we catch function deprecations early.

Bill-Sommerfeld committed:
Revise rbac patch to set library dependencies in configure.ac

Fix all compile-time warnings.

Move illumos additions in configure.ac before the
AC_SUBST([MODULE_BLOCK]) which is described as "must come after last
PY_STDLIB_MOD()".

Build rbac with -Werror so we catch function deprecations early.

Bill-Sommerfeld committed:
Revise 20-test_fixes to avoid test failures.

Avoid looking in the wrong /usr/lib/python for the pkg package.

Add missing import sys for test_gdb.

Bill-Sommerfeld committed:
In 37-test-zipfile.patch, capitalize OSError correctly.

Bill-Sommerfeld committed:
Work around truncated timestamp issue in test_tarfile

Bill-Sommerfeld committed:
Solaris's libc timezone code is returning EET rather than MST for
times in Europe/Kyiv in 1984.  Unclear why this is but it's not a
python issue.

Bill-Sommerfeld committed:
weaken assertions in test_shutil to tolerate rather than expect an exception

The test_shutil test expects an error if you sendfile a file to itself.

Our code apparently doesn't do that but it also doesn't corrupt the
file; weaken the assertions accordingly.

Bill-Sommerfeld committed:
two tests of test_eintr hang on Illumos; they are skipped on macos for that reason

Bill-Sommerfeld committed:
Tolerate systems (like illumos) where tcgetwinsize fails if the window size is unset.

Most systems return zero for window size on a fresh pty, but illumos
fails the call with EINVAL until at least one of the fields is set to
a nonzero value.

Bill-Sommerfeld committed:
Revert to OpenSSL 1.1 due to test failures.

A bunch of unit tests (including test_imaplib and several asyncio
tests) that use ssl started failing after I switched to OpenSSL 3.1.
Revert to 1.1 in the short term to keep tests clean and avoid
inflicting broken SSL libraries on Python programs.

When the problem is root caused and fixed we can move forward again.

Bill-Sommerfeld committed:
Remove stale ac_cv_* settings.

We disable epoll via a patch so we don't also need to disable it in
the Makefile.

Illumos now supports getrandom() which Python prefers over
getentropy(); we no longer need to disable the getentropy probe
because Python will never use it.

Bill-Sommerfeld committed:
All tests either passing or skipped except for test_dtrace

(as an aside, test_dtrace is doing something odd with readelf that I
think bears no relation to how dtrace is implemented on Illumos)

Bill-Sommerfeld committed:
Checkpoint my TODO file

Bill-Sommerfeld committed:
Python 3.12.2

mtelka commented on components/python/python-312/pkg5:
```diff
@@ -9,7 +9,7 @@
         "library/ncurses",
         "library/readline",
-        "library/security/openssl-31",
+        "library/security/openssl-11",
```
Something is broken with openssl. :-(

mtelka replied:
Ah, you switched back to 1.1.  That explains this change.

Bill-Sommerfeld replied:
Yes.   When I get a chance I'll dig into the test failures but until then, working 1.1 is better than broken 3.1...

mtelka replied:
Yes, sure. Thanks.

mtelka commented on components/python/python-312/patches/44-disable_epoll.patch:
```diff
@@ -0,0 +1,22 @@
+
+Without this patch, python detects and uses epoll which only exists in
+OmniOS for lx zones and Linux compatibility. It is not quite the same as
+the Linux implementation and can cause socket related failures in python.
```
This is something that needs some love:

- we are not OmniOS :-)
- we do not have lx zones
- it is strange that OmniOS claims (this patch was apparently copied from OmniOS) that since epoll is for Linux compatibility then it should not be used.  And the reason is that our implementation is not quite the same as on Linux.  The [epoll support](https://www.illumos.org/issues/5640) was added to be fully compatible with Linux and to make porting of software from Linux easier.  If there is some notable difference, then it is a bug in illumos that should be (at least) reported, and then investigated and eventually fixed.
- it should be checked/tested if this patch is still needed;  if it is still needed then it means that something in illumos epoll does not work as expected (see above), or there is some bug in Python itself.  Either way we need to report this as a bug.

Thanks.

alarcher replied:
The epoll support was not intended to be fully compatible as some behaviours which were considered as buggy in the Linux implementation were not replicated (there was a presentation by Cantrill along these lines), therefore illumos epoll can only be used reliably in a subset of the situations encountered. I cannot remember the details but there are situations where the Linux epoll "works" which should be considered as undefined behaviour, like referencing resources that are long gone. Mimicing Linux's behaviour would rely on border effects. In such cases disabling epoll was the only reasonable way to go. The good news is that usually epoll would just hang on illumos when such edge cases are met.

alarcher replied:
For the same reason epoll is disabled in Xorg and other graphical toolkits.

mtelka replied:
Okay, understood.  But still we need better comment here in this patch.  Reference to more info would be great.  For exactly the reason you stated: "I cannot remember..."

Also, hang in epoll (is the hang in illumos or in the epoll consumer?) cannot be considered as good news.  It sounds like a bug.

alarcher replied:
The hang was in the consumer and we were told that the use case was of one of those not covered by the illumos implementation because the Linux implementation only "works" because of a border effect.
Sorry that I cannot remember better, we had a thread a few years ago but a lot has happened since then.

mtelka replied:
It looks like there are many problems related to epoll in Python itself.

Examples:
- on Linux: https://github.com/python/cpython/issues/112741
- ~on Windows (yes, they use epoll even on Windows!): https://github.com/python/cpython/issues/111604~ (EDIT: event loop is implemented differently there)

But I found only one (long time ago solved) issue reported there related to illumos and epoll:
- https://github.com/python/cpython/issues/73572

So we definitely need more bug reports :-).

mtelka replied:
It looks like the [epoll(7) man page](https://illumos.org/man/7/epoll) contains all needed information, including caveats about Linux epoll implementation.

mtelka committed:
Drop _LARGEFILE64_SOURCE; we compile 64-bit only

mtelka commented:
I think it would be better to mark this as draft so it is not merged accidentally :-).

mtelka commented:
Once #16009 is merged please rebase on top of it and generate new sample-manifest.

mtelka committed:
Rework epoll patch and improve the description.

mtelka committed:
use parallel build

mtelka committed:
06-rbac.patch: fix SyntaxWarning

mtelka committed:
simplify comment for 35-20142.patch

mtelka committed:
no test redirect

Bill-Sommerfeld committed:
Add patch description.

Bill-Sommerfeld committed:
Regenerate sample-manifest after #16009 and adjust real manifests to match

Bill-Sommerfeld commented:
> Once #16009 is merged please rebase on top of it and generate new sample-manifest.

Done.

Bill-Sommerfeld committed:
Restore python function boundary dtrace probes (WIP)

mtelka committed:
Use default manifest template

mtelka committed:
python.p5m: use transform for autopyc=false; add missing smtpd.py

mtelka committed:
Create symlinks in post-install

Bill-Sommerfeld committed:
Attempt to restore dtrace line probe (incomplete WIP)

Bill-Sommerfeld committed:
checkpoint TODO again

mtelka committed:
Set mediators using transforms

mtelka committed:
Fix python-config symlink and libpython312.so

mtelka committed:
Improve 64-bit object in 32-bit path lint error handling

mtelka committed:
Rename python.p5m to Python.p5m and remove commented out actions

mtelka committed:
$(PYTHON.$(PYTHON_VERSION).$(BITS)) -> $(PYTHON)

Bill-Sommerfeld committed:
Enable USE_CTF

Bill-Sommerfeld committed:
Add GCC_GNU_SHORT_TRIPLET for packages that use ...-solaris2 instead of ...-solaris2.11

Bill-Sommerfeld committed:
Pull in upstream fix for test_xml_etree and test_xml_etree_c

Bill-Sommerfeld committed:
Fix recognition of target triple for illumos.

From https://github.com/omniosorg/omnios-build/pull/3492

Bill-Sommerfeld reviewed (commented):
I've submitted [9e956fd](https://github.com/OpenIndiana/oi-userland/pull/15955/commits/9e956fd68ba0541e20a41e55c47065027d7dbe94) as a separate pull request: OpenIndiana/oi-userland#16200

AndWac commented:
Any news here?

AndWac commented:
Superseded by #22983
